dwh/indicator_lib.py

```diff
@@ -120,4 +120,73 @@ def calc_TX_NEW(patient_agg_obs:  pd.DataFrame, ARV_plan: str,
+                start_drug: List[str], end_date_str: str = None) -> pd.DataFrame:
+  """Calculates TX_CURR indicator with its corresponding disaggregations.
+
+  TX_CURR indicator counts the number of adults and children currently receiving
+  antiretroviral therapy (ART).
+
+  Args:
+    patient_agg_obs: A DataFrame generated by `patient_query.find_patient_aggregates()`.
+    ARV_plan: The concept question code for ANTIRETROVIRAL PLAN
+    start_drug: The concept answer codes for START DRUG, CONTINUE REGIMEN, REFILLED
+    end_date_str: The string representation of the last date as 'YYYY-MM-DD'.
+  Returns:
+    The aggregated DataFrame with age/gender buckets.
+  """
+  end_date = datetime.today()
+  if end_date_str:
+    end_date = parse_date(end_date_str)
+  temp_df = patient_agg_obs[(patient_agg_obs['code'] == ARV_plan)].copy()
+  temp_df['TX_CURR'] = (temp_df['last_value_code'].isin(start_drug))
+  temp_df = _gen_counts_and_ratio(temp_df, end_date, 'TX_CURR')
+  return temp_df
+
+def calc_TB_STAT(patient_agg_obs:  pd.DataFrame, TB_TX_plan: str,
+                 ARV_plan: str,TB_test_status: str,start_drug: List[str],
+                 yes_code, end_date_str: str = None) -> pd.DataFrame:
+  """Calculates TB_STAT indicator with its corresponding disaggregations.
+
+  TB_STAT indicator counts the number of new and relapse TB cases with
+  documented HIV status
+
+  Args:
+    patient_agg_obs: A DataFrame generated by `patient_query.find_patient_aggregates()`.
+    TB_TX_plan: The concept question code for TB treatment PLAN
+    ARV_plan: The concept question code for ANTIRETROVIRAL PLAN
+    TB_test_status: The concept question code for TB test done (Yes /No)
+    start_drug: The concept answer codes for START DRUG
+    yes_code: The concept answer codes for YES
+    end_date_str: The string representation of the last date as 'YYYY-MM-DD'.
+  Returns:
+    The aggregated DataFrame with age/gender buckets.
+  """
+  end_date = datetime.today()
+  if end_date_str:
+    end_date = parse_date(end_date_str)
+  # Check for TB TREATMENT PLAN (if START it means diagnosis was done)
+  tb_tx_df = patient_agg_obs[(patient_agg_obs['code'] == TB_TX_plan)].copy()
+  tb_tx_df['TB_TX'] = (tb_tx_df['last_value_code'].isin(start_drug))
+
+  # check if TB was diagnosed
+  tb_test_df = patient_agg_obs[(patient_agg_obs['code'] == TB_test_status)].copy()
+  tb_test_df['TB_test'] = (tb_test_df['last_value_code'] == yes_code)
+
+  # check if patient is HIV +
+  art_tx_df = patient_agg_obs[(patient_agg_obs['code'] == ARV_plan)].copy()
+  art_tx_df['HIV_positive'] = (art_tx_df['last_value_code'].notnull())
+
+  # TODO: Need a simpler way of doing multi-code evaluation instead of the above
+  # join the 3 DF
+  temp_df=tb_tx_df.merge(tb_test_df[['patientId', 'TB_test']],
+                         on='patientId').merge(art_tx_df[['patientId', 'HIV_positive']],
+                                               on='patientId')
+  # evaluate
```

Comment:
@bashir2  how can we simplify the ability to  do the above.,

Reply:
I actually think this is an okay way of doing multi-code evaluation; I mean I expect this join to be fairly efficient because of the equality join condition on `patientId`. Are you experiencing (or have concerns about) efficiency issues?

Reply:
Not performance really, but complexity of indicator definition for non-pepfar metrics. A good example would be [tb_screening results ](https://github.com/AMPATH/etl/blob/2683f94708cda320c262a8daa02138f2e443af47/etl-scripts/stored-procedures/generate_hiv_summary_v15_9.sql#L1158-L1182) - will require over 6 df evaluations and joins

Since there are few of these indicators, let's worry about this later

Thanks @bashir2